### ISSUE TITLE
[Bug]: reference-only modes fail with s_min_uncond > 0

### ISSUE THREAD

**Issue by asagi4** [Mikubill/sd-webui-controlnet#1380]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

When Negative conditioning minimum sigma is enabled, image generation proceeds as normal first and then fails with an exception.

This seems to happen with most samplers, but not with DDIM or UniPC for some reason. I didn't test all of them

The exact error seems to differ depending on which optimization (or none) is enabled, so `--opt-sdp-no-mem-attention` gives a different exception than `--xformers` etc.

### Steps to reproduce the problem

1. Enable negative conditioning minimum sigma; any value should do, but higher will trigger the bug earlier. (s_min_uncond, under "Optimizations")
2. Enable any of the "reference" ControlNet modes with any reference image
3. Generate anything with eg. the "Euler a" sampler
4. Observe failure after a few steps

### What should have happened?

It shouldn't fail, or should at least inform the user that s_min_uncond > 0 is not supported.

### Commit where the problem happens

webui: 38a2324dc3033125a2b200cb1fca52bf361e0046 (dev branch, but should happen on master too)
controlnet: 06515a669d4e95fd39a6fc68df1021ec50643002

### What browsers do you use to access the UI ?

Mozilla Firefox

### Command Line Arguments

```Shell
No
```

### List of enabled extensions

I disabled everything except controlnet

### Console logs

```Shell
Example exception:
Traceback (most recent call last):
  File "/home/sd/git/stable-diffusion-webui/modules/call_queue.py", line 57, in f
    res = list(func(*args, **kwargs))
  File "/home/sd/git/stable-diffusion-webui/modules/call_queue.py", line 37, in f
    res = func(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/modules/txt2img.py", line 57, in txt2img
    processed = processing.process_images(p)
  File "/home/sd/git/stable-diffusion-webui/modules/processing.py", line 611, in process_images
    res = process_images_inner(p)
  File "/home/sd/git/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/batch_hijack.py", line 42, in processing_process_images_hijack
    return getattr(processing, '__controlnet_original_process_images_inner')(p, *args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/modules/processing.py", line 729, in process_images_inner
    samples_ddim = p.sample(conditioning=p.c, unconditional_conditioning=p.uc, seeds=p.seeds, subseeds=p.subseeds, subseed_strength=p.subseed_strength, prompts=p.prompts)
  File "/home/sd/git/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/hook.py", line 224, in process_sample
    return process.sample_before_CN_hack(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/modules/processing.py", line 977, in sample
    samples = self.sampler.sample(self, x, conditioning, unconditional_conditioning, image_conditioning=self.txt2img_image_conditioning(x))
  File "/home/sd/git/stable-diffusion-webui/modules/sd_samplers_kdiffusion.py", line 383, in sample
    samples = self.launch_sampling(steps, lambda: self.func(self.model_wrap_cfg, x, extra_args={
  File "/home/sd/git/stable-diffusion-webui/modules/sd_samplers_kdiffusion.py", line 257, in launch_sampling
    return func()
  File "/home/sd/git/stable-diffusion-webui/modules/sd_samplers_kdiffusion.py", line 383, in <lambda>
    samples = self.launch_sampling(steps, lambda: self.func(self.model_wrap_cfg, x, extra_args={
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/k-diffusion/k_diffusion/sampling.py", line 145, in sample_euler_ancestral
    denoised = model(x, sigmas[i] * s_in, **extra_args)
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/modules/sd_samplers_kdiffusion.py", line 137, in forward
    x_out = self.inner_model(x_in, sigma_in, cond=make_condition_dict([cond_in], image_cond_in))
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/k-diffusion/k_diffusion/external.py", line 112, in forward
    eps = self.get_eps(input * c_in, self.sigma_to_t(sigma), **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/k-diffusion/k_diffusion/external.py", line 138, in get_eps
    return self.inner_model.apply_model(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/modules/sd_hijack_utils.py", line 17, in <lambda>
    setattr(resolved_obj, func_path[-1], lambda *args, **kwargs: self(*args, **kwargs))
  File "/home/sd/git/stable-diffusion-webui/modules/sd_hijack_utils.py", line 28, in __call__
    return self.__orig_func(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/models/diffusion/ddpm.py", line 858, in apply_model
    x_recon = self.model(x_noisy, t, **cond)
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/models/diffusion/ddpm.py", line 1335, in forward
    out = self.diffusion_model(x, t, context=cc)
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1538, in _call_impl
    result = forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/hook.py", line 453, in forward_webui
    return forward(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/hook.py", line 409, in forward
    outer.original_forward(
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/modules/diffusionmodules/openaimodel.py", line 797, in forward
    h = module(h, emb, context)
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/modules/diffusionmodules/openaimodel.py", line 84, in forward
    x = layer(x, context)
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/modules/attention.py", line 334, in forward
    x = block(x, context=context[i])
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/modules/attention.py", line 269, in forward
    return checkpoint(self._forward, (x, context), self.parameters(), self.checkpoint)
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/modules/diffusionmodules/util.py", line 121, in checkpoint
    return CheckpointFunction.apply(func, len(inputs), *args)
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/home/sd/git/stable-diffusion-webui/repositories/stable-diffusion-stability-ai/ldm/modules/diffusionmodules/util.py", line 136, in forward
    output_tensors = ctx.run_function(*ctx.input_tensors)
  File "/home/sd/git/stable-diffusion-webui/extensions/sd-webui-controlnet/scripts/hook.py", line 489, in hacked_basic_transformer_inner_forward
    x = self.attn2(self.norm2(x), context=context) + x
  File "/home/sd/git/stable-diffusion-webui/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/sd/git/stable-diffusion-webui/modules/sd_hijack_optimizations.py", line 515, in scaled_dot_product_no_mem_attention_forward
    return scaled_dot_product_attention_forward(self, x, context, mask)
  File "/home/sd/git/stable-diffusion-webui/modules/sd_hijack_optimizations.py", line 490, in scaled_dot_product_attention_forward
    k = k_in.view(batch_size, -1, h, head_dim).transpose(1, 2)
RuntimeError: shape '[2, -1, 8, 40]' is invalid for input of size 24640
```

### Additional information

I'm running SD on RHEL 9 with a NVIDIA GeForce RTX 3060 under Python 3.10.11
torch is 2.0.0+cu118

This is not a particularly new bug, I just didn't get around to reporting it before now.

**Comment by lllyasviel** [comment]
what is Negative conditioning minimum sigma and how it works?

**Comment by lllyasviel** [comment]
i tried to understand but really have no idea. but i think we expect cn to support as much as possible a1111 features so I will add help wanted flag.

**Comment by Dekker3D** [comment]
The gist of it is described in this PR: https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/9177

Basically, the CFG thing involves calculating a conditioned and unconditioned output, taking the difference and scaling it by CFG. This PR sets CFG to 1 and/or just skips calculating the unconditioned output, but does that only for the negative prompt part, and only at "low sigma values" where the threshold is determined by that setting.

I have no clue how the "sigma value" is calculated.

**Comment by rkiga** [comment]
>This seems to happen with most samplers, but not with DDIM or UniPC for some reason. I didn't test all of them

"Negative Guidance minimum sigma" only works with the k-diffusion samplers. So PLMS shouldn't be affected by this conflict either.

> I have no clue how the "sigma value" is calculated.

TLDR: sigma is the amount of noise for any given step, and the default max sigma (at the end of training or the start of inference) = 14.6146

If you look at the image of the noise schedule ramp below, more than 85% of denoising is done in the first half of steps, because the first few steps do a lot of rough denoising very quickly. So if you set "Negative Guidance minimum sigma" to 15% of 14.6146, which = ~2.2 then "Negative Guidance minimum sigma" will trigger after ~half of your total steps are complete. And then alternating steps will skip the negative prompt.

---

I'm still trying to learn how it all works, but this is my current understanding:

Stable Diffusion has 1000 acceptable sigmas, spanning from 0.0292 to 14.6146, which Birch-san listed: https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/b34b25b4c941819d34f29be6c4c1ec01e64585b4 

_sigma_ here is the amount of noise on the latent image during any given step. And 14.6146 is the "maximum noise it would have seen during training" https://forums.fast.ai/t/init-noise-sigma/101423 
"The model was trained to associate a particular amount of noise with a particular value of _t_." Where _t_ = timestep = step. 

During inference, if you choose 16 steps, the sampler (which is sometimes called the scheduler) sets a noise schedule that ramps down like this: https://stable-diffusion-art.com/samplers/#Karras_noise_schedule (each unit on the y-axis is 2 sigma)

Those sigma values are rounded to the nearest of the 1000 acceptable values. So the sampler starts with the max noisy latent, then denoises until it gets to the next sigma value on the schedule, and then starts the next step.... until the end.

Looking at the image of the schedule above, if you set your "Negative Guidance minimum sigma" to 1.0 and set your steps to 16, then sigma will go below that 1.0 threshold just before step 10. So it ignores your negative prompt for steps 10, 12, 14, and 16, which means fewer calculations are needed. 

But I have no idea why this breaks ControlNet reference-only. The PR#9177, linked above, says it breaks Regional Prompter too, or did in the past.

**Comment by DominikDoom** [comment]
I was not able to resolve it myself, but want to bump it and give some additional info on what is going on here:

This breaks reference-only due to the custom attention injection in **hook.py**, and since the reference ones are the only modules using that, the rest works fine. From what I've seen it breaks at two points (possibly more since I haven't gotten it to run further):

1. When calling `ref_xt = predict_q_sample(...)` for batch sizes > 1
   - My guess is that this happens because the webui sets `x_in = x_in[:-batch_size]` in **sd_samplers_cfg_denoise.py** if s_min_uncond is > 0, which I interpret as cutting off the unneeded half for steps where uncond is skipped.
   - The error will also always mention "The size of tensor a (&lt;batch size&gt;) must match the size of tensor b (2*&lt;batch size&gt;) at non-singleton dimension 0".
2. For batch size 1, during the `hacked_basic_transformer_inner_forward(...)` call, specifically at the end:
   - `x = self.attn2(self.norm2(x), context=context) + x`
   - For this, it will only occur once the sigma threshhold is reached (in my case with s_min_uncond = 2.5 at step 12/28)
   - During experimentation, by passing s_min_uncond=0.0 as a parameter in the attn2 call, the error would only happen in the next line for the ff call, where I tried the same without success.

I lack the knowledge about the inner workings of controlnet / the attention injection functions, but I guess this should be resolveable by either replicating the skip code from **sd_samplers_cfg_denoise.py**, which mainly replaces uncond with cond so the shape is correct from what I could gather, or somehow accounting for tensor sizes changing between steps in the affected custom forward calls.
